sdk/program/src/poseidon.rs

```diff
@@ -0,0 +1,411 @@
+//! Hashing with the [Poseidon] hash function.
+//!
+//! [Poseidon]: https://www.poseidon-hash.info/
+
+use thiserror::Error;
+
+/// Length of Poseidon hash result.
+pub const HASH_BYTES: usize = 32;
+
+#[derive(Error, Debug)]
+pub enum PoseidonSyscallError {
+    #[error("Invalid curve.")]
+    InvalidCurve,
+    #[error("Invalid endianness.")]
+    InvalidEndianness,
+    #[error("Invalid number of inputs. Maximum allowed is 12.")]
+    InvalidNumberOfInputs,
+    #[error("Input is an empty slice.")]
+    EmptyInput,
+    #[error(
+        "Invalid length of the input. The length matching the modulus of the prime field is 32."
+    )]
+    InvalidInputLength,
+    #[error("Input is larger than the modulus of the prime field.")]
+    InputLargerThanModulus,
+    #[error("Failed to convert a vector of bytes into an array.")]
+    VecToArray,
+    #[error("Failed to convert the number of inputs from u64 to u8.")]
+    U64Tou8,
+    #[error("Invalid width. Choose a width between 2 and 16 for 1 to 15 inputs.")]
+    InvalidWidthCircom,
+    #[error("Unexpected error")]
+    Unexpected,
+}
+
+impl From<u64> for PoseidonSyscallError {
+    fn from(error: u64) -> Self {
+        match error {
+            1 => PoseidonSyscallError::InvalidCurve,
+            2 => PoseidonSyscallError::InvalidEndianness,
+            3 => PoseidonSyscallError::InvalidNumberOfInputs,
+            4 => PoseidonSyscallError::EmptyInput,
+            5 => PoseidonSyscallError::InvalidInputLength,
+            6 => PoseidonSyscallError::InputLargerThanModulus,
+            7 => PoseidonSyscallError::VecToArray,
+            8 => PoseidonSyscallError::U64Tou8,
+            9 => PoseidonSyscallError::InvalidWidthCircom,
+            _ => PoseidonSyscallError::Unexpected,
+        }
+    }
+}
+
+impl From<PoseidonSyscallError> for u64 {
+    fn from(error: PoseidonSyscallError) -> Self {
+        match error {
+            PoseidonSyscallError::InvalidCurve => 1,
+            PoseidonSyscallError::InvalidEndianness => 2,
+            PoseidonSyscallError::InvalidNumberOfInputs => 3,
+            PoseidonSyscallError::EmptyInput => 4,
+            PoseidonSyscallError::InvalidInputLength => 5,
+            PoseidonSyscallError::InputLargerThanModulus => 6,
+            PoseidonSyscallError::VecToArray => 7,
+            PoseidonSyscallError::U64Tou8 => 8,
+            PoseidonSyscallError::InvalidWidthCircom => 9,
+            PoseidonSyscallError::Unexpected => 10,
+        }
+    }
+}
+
+/// Eliptic curve.
+#[repr(u64)]
+pub enum Curve {
```

Comment:
Sorry, I totally forgot about that!

Now it's done. I also renamed the enum to `Parameters`, which I think is more accurate, especially in case we want to use the same elliptic curve types with different S-Boxes in the future. And I added detailed docstrings.